File: admin/src/shared/api/admin-observability.ts
// 作者: 杨永的Agent
// 日期: 2026-05-18
// 修改功能: **`prompt.snapshot`** / **`llm.chat.faq`** / **`llm.{read.market.*|read.account.balance|wealth.holdings_read}`**（`gateway*` / `stepKind` · FE_HANDOFF 0014～0016）；现货撤单 **`cancel_order`** stepKind（FE_HANDOFF 2026-05-18）
// 作者: 杨永的Agent
// 日期: 2026-05-15
// 修改功能: `summary` 可选 **`venue` / `canonicalOp`**（ADR-004 · FE_HANDOFF）
// 日期: 2026-05-14
// 修改功能: `orderRequest.limitOrderMeta`（限价 base / price / TIF · trade.spot.limit_order）
// 日期: 2026-05-14
// 修改功能: `orderRequest.flashMarketMeta`（volume 语义协查 · MARKET BUY/SELL）
// 日期: 2026-05-14
// 修改功能: `summary.exchangeReadPreviewSummary` / `appErrorCode`（只读路由 §5.2.2）
// 日期: 2026-05-27
// 修改功能: 执行详情 Tab **`/queue|events|retries|recovery`** GET 与类型（FE_HANDOFF 2026-05-27）
// 日期: 2026-05-27
// 修改功能: **`promptBindingResolved` / `skillSpecRead`** 时间线 summary；**`getScenarioSkillScope`**（FE_HANDOFF 2026-05-26）
// 日期: 2026-05-14
// 修改功能: `ObservabilityTimelineSummary`（transitionTrigger/stepKind 等 · §5.2）
// 日期: 2026-05-14
// 修改功能: `GET …/executions/{id}/timeline`（ObservabilityTimelineResponse · FE_HANDOFF FR-MC801）
// 日期: 2026-05-13
// 修改功能: Admin Observability — `GET`/`DELETE …/executions`（agent_execution，FE_HANDOFF）

import { HTTPError } from 'ky'

import { AppError } from '@/shared/api/errors'
import { httpClient, toAppError } from '@/shared/api/http-client'
import type { ResolvedPromptBinding } from '@/shared/api/admin-prompt-packs'

/** 时间线 `agent.skill.spec_read` · `summary.skillSpecRead` 镜像 */
export interface ObservabilitySkillSpecReadSummary {
  skillId?: string | null
  skillSpecVersion?: string | null
  specDigest?: string | null
  phase?: string | null
}

/** 场景 Skill 范围 · `GET …/scenarios/{scenarioId}/skill-scope` */
export interface ScenarioSkillScopeSkillItem {
  skillId: string
  skillSpecVersion?: string | null
  specDigest?: string | null
  contractComplete?: boolean | null
  role?: string | null
}

export interface ScenarioSkillScopeResponse {
  scenarioId: string
  category: string
  mode: 'write_skill' | 'read_only' | 'unmapped_write' | string
  skills: ScenarioSkillScopeSkillItem[]
  promptStrategyPackId?: string | null
  promptStrategyVersion?: string | null
  promptSkillScopeRef?: string | null
  narrative: string
}

const PREFIX = 'v1/admin/observability'

/** 与 `AdminAgentExecutionItem`（camelCase）一致 */
export interface AdminAgentExecutionItem {
  executionId: string
  userId: string
  scenarioId: string | null
  channel: string | null
  state: string
  source: string | null
  idempotencyKey: string | null
  note: string | null
  /** 已发布 Prompt 包版本（accept 时未手填可由服务端写入；无发布包时为 `null` / 省略） */
  promptPackVersion?: string | null
  /** 与 effective prompts 同源绑定快照 */
  resolvedPromptBinding?: ResolvedPromptBinding | Record<string, unknown> | null
  createdAt: string
  updatedAt: string | null
}

export interface AdminAgentExecutionListResponse {
  items: AdminAgentExecutionItem[]
  total: number
}

/** `orderRequest.flashMarketMeta` · 标明 MARKET 单 wire `volume` 与用户意向的区别 */
export interface ObservabilityFlashMarketMeta {
  volumeSemantics?: string
  baseQtyUserRequested?: string | number
  /** BUY：`volume` 字段实为计价金额，与本字段一致 */
  quoteAmountOnWire?: string | number
  /** SELL：`volume` 与 base 数量一致时常与本字段对齐 */
  baseQtyOnWire?: string | number
  lastPriceUsed?: string | number
  volumeSemanticsNote?: string
  userVolumeInput?: string
}

/** `orderRequest.limitOrderMeta` · LIMIT 协查（与闪兑 `flashMarketMeta` 并列） */
export interface ObservabilityLimitOrderMeta {
  volumeSemantics?: string
  limitPriceUserRequested?: string | number
  baseQtyUserRequested?: string | number
  timeInForce?: string
}

/** 时间线里脱敏的下单请求快照（嵌于 `summary.orderRequest`） */
export interface ObservabilityOrderRequestSnapshot {
  path?: string
  symbol?: string
  side?: string
  type?: string
  volume?: string | number
  price?: string | number
  newClientOrderId?: string
  flashMarketMeta?: ObservabilityFlashMarketMeta
  limitOrderMeta?: ObservabilityLimitOrderMeta
}

/**
 * 时间线 `summary` 常见键（详见 `API_ADMIN_OBSERVABILITY_EXECUTIONS.md` §5.2）。
 * 服务端仍可追加任意字段，故保留索引签名由调用方收窄读取。
 */
export type ObservabilityTimelineSummary = Record<string, unknown> & {
  /** 如 flash：`quote` / `confirm_prompt` / `confirm_accept` / `submit_order` */
  stepKind?: string
  /** 状态迁移触发（若后端写入） */
  transitionTrigger?: string
  scenarioId?: string
  outcome?: string
  methodPathSummary?: string
  exchangeOutcome?: string
  clientOrderRef?: string
  httpStatus?: number
  orderRequest?: ObservabilityOrderRequestSnapshot
  exchangeResponsePreview?: Record<string, unknown>
  /** 只读路由成功：`routing/execute` · Telegram `read.*`（§5.2.2） */
  exchangeReadPreviewSummary?: Record<string, unknown> | string | number | boolean | null
  /** 只读失败：`routing.read.failed` */
  appErrorCode?: string
  /** 交易场所（限价/闪兑写路径摘要 · ADR-004） */
  venue?: string
  /** 规范操作名（如 `place_order`） */
  canonicalOp?: string
  /** `prompt.snapshot` / `llm.chat.faq` / 托管预览 `llm.*` 叙述：**Prompt 版本与网关快照** */
  promptPackVersion?: string | null
  /** 与 effective 同源绑定 */
  resolvedPromptBinding?: ResolvedPromptBinding | Record<string, unknown> | null
  /** `agent.prompt.binding_resolved` 镜像（与 `resolvedPromptBinding` 同形） */
  promptBindingResolved?: ResolvedPromptBinding | Record<string, unknown> | null
  /** `agent.skill.spec_read` 镜像 */
  skillSpecRead?: ObservabilitySkillSpecReadSummary | Record<string, unknown> | null
  skillId?: string | null
  skillSpecVersion?: string | null
  specDigest?: string | null
  /** 闲聊网关：目录模型 id（已解析时） */
  gatewayModelId?: string | null
  /** 对端 HTTP model 线 id（apiModel ?? modelId） */
  gatewayUpstreamModel?: string | null
  gatewayProviderId?: string | null
  /** 方舟 `/api/v3/responses` 路径 */
  useArkProtocol?: boolean | null
}

/** `/executions/{id}/timeline` 单项 · camelCase JSON */
export interface ObservabilityTimelineItem {
  ts: string
  eventName: string
  executionId?: string | null
  userId?: string | null
  summary: ObservabilityTimelineSummary
}

export interface ObservabilityTimelineResponse {
  items: ObservabilityTimelineItem[]
}

export type ListAdminExecutionsParams = {
  limit?: number
  offset?: number
  executionId?: string
  userId?: string
  channel?: string
  scenarioId?: string
  state?: string
  keyword?: string
  createdAfter?: string
  createdBefore?: string
}

async function normalizeKyError(error: unknown): Promise<never> {
  if (error instanceof HTTPError) {
    const cloned = error.response.clone()
    const body: unknown = await cloned.json().catch(() => null)
    if (
      body !== null &&
      typeof body === 'object' &&
      'detail' in body &&
      typeof (body as { detail: unknown }).detail === 'string'
    ) {
      const code =
        'code' in body && typeof (body as { code: unknown }).code === 'string'
          ? (body as { code: string }).code
          : undefined
      throw new AppError((body as { detail: string }).detail, {
        status: error.response.status,
        code,
      })
    }
    if (
      body !== null &&
      typeof body === 'object' &&
      'message' in body &&
      typeof (body as { message: unknown }).message === 'string'
    ) {
      const code =
        'code' in body && typeof (body as { code: unknown }).code === 'string'
          ? (body as { code: string }).code
          : undefined
      throw new AppError((body as { message: string }).message, {
        status: error.response.status,
        code,
      })
    }
  }
  throw await toAppError(error)
}

function optionalSearchParams(p: ListAdminExecutionsParams): Record<string, string> {
  const out: Record<string, string> = {}
  const lim = p.limit ?? 50
  const off = p.offset ?? 0
  out.limit = String(Math.min(200, Math.max(1, lim)))
  out.offset = String(Math.max(0, off))
  const eid = p.executionId?.trim()
  if (eid) out.executionId = eid
  const uid = p.userId?.trim()
  if (uid) out.userId = uid
  const ch = p.channel?.trim()
  if (ch) out.channel = ch
  const sid = p.scenarioId?.trim()
  if (sid) out.scenarioId = sid
  const st = p.state?.trim()
  if (st) out.state = st
  const kw = p.keyword?.trim()
  if (kw) out.keyword = kw
  const ca = p.createdAfter?.trim()
  if (ca) out.createdAfter = ca
  const cb = p.createdBefore?.trim()
  if (cb) out.createdBefore = cb
  return out
}

export async function listAdminObservabilityExecutions(
  params: ListAdminExecutionsParams = {},
): Promise<AdminAgentExecutionListResponse> {
  try {
    return await httpClient
      .get(`${PREFIX}/executions`, { searchParams: optionalSearchParams(params) })
      .json<AdminAgentExecutionListResponse>()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

export async function getAdminObservabilityExecutionTimeline(
  executionId: string,
): Promise<ObservabilityTimelineResponse> {
  const id = executionId.trim()
  if (!id) throw new AppError('executionId 不可为空', { status: 400 })
  try {
    return await httpClient
      .get(`${PREFIX}/executions/${encodeURIComponent(id)}/timeline`)
      .json<ObservabilityTimelineResponse>()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

export async function getAdminObservabilityExecution(executionId: string): Promise<AdminAgentExecutionItem> {
  const id = executionId.trim()
  if (!id) throw new AppError('executionId 不可为空', { status: 400 })
  try {
    return await httpClient
      .get(`${PREFIX}/executions/${encodeURIComponent(id)}`)
      .json<AdminAgentExecutionItem>()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

/** 硬删除持久化 execution · 成功 **204** 无 Body */
export async function deleteAdminObservabilityExecution(executionId: string): Promise<void> {
  const id = executionId.trim()
  if (!id) throw new AppError('executionId 不可为空', { status: 400 })
  try {
    await httpClient.delete(`${PREFIX}/executions/${encodeURIComponent(id)}`)
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

/** FR-MC803 · `GET …/executions/{id}/tool-calls` */
export interface ObservabilityToolCallItem {
  toolId: string
  toolCallSeq: number
  invocationState: string
  phase?: string | null
  durationMs?: number | null
  agentSubAccountId?: string | null
  eventName?: string | null
  stepKind?: string | null
  ts?: string | null
  summary?: Record<string, unknown>
}

export interface ObservabilityToolCallsResponse {
  items: ObservabilityToolCallItem[]
}

/** FR-MC804 · `GET …/executions/{id}/llm`（无 messages 全文） */
export interface ObservabilityLlmCallItem {
  eventName: string
  toolCallSeq: number
  ts: string
  modelId?: string | null
  gatewayModelId?: string | null
  gatewayUpstreamModel?: string | null
  gatewayProviderId?: string | null
  outcome?: string | null
  stepKind?: string | null
  scenarioId?: string | null
  promptPackVersion?: string | null
  inputTokens?: number | null
  outputTokens?: number | null
  totalTokens?: number | null
}

export interface ObservabilityLlmUsageResponse {
  modelId?: string | null
  inputTokens?: number | null
  outputTokens?: number | null
  totalTokens?: number | null
  calls: ObservabilityLlmCallItem[]
}

/** `GET …/executions/{id}/queue` */
export interface ExecutionTaskQueueItem {
  taskId: string
  executionId: string
  state: 'PENDING' | 'RUNNING' | 'BLOCKED' | string
  scheduledAt: string
  stepKey?: string | null
  stepLabelZh?: string | null
}

export interface ExecutionTaskQueueResponse {
  items: ExecutionTaskQueueItem[]
}

/** `GET …/executions/{id}/events` */
export interface ExecutionRuntimeEventItem {
  executionId: string
  at: string
  eventType: string
  summary: string
  seq?: number | null
}

export interface ExecutionRuntimeEventsResponse {
  items: ExecutionRuntimeEventItem[]
}

/** `GET …/executions/{id}/retries` */
export interface ExecutionRetryItem {
  attempt: number
  at: string
  reason: string
  eventType?: string | null
  seq?: number | null
}

export interface ExecutionRetriesResponse {
  retryCount: number
  items: ExecutionRetryItem[]
}

/** `GET …/executions/{id}/recovery` */
export interface ExecutionRecoveryResponse {
  executionId: string
  scenarioId?: string | null
  executionState: string
  caseKind: string
  resolutionStatus: string
  stillUnknown: boolean
  lastReconcileAt?: string | null
  lastReconcileAtSeq?: number | null
  reconcileId?: string | null
  userMessage?: string | null
  hints?: string[]
  recoveryTitle: string
  recoveryBody: string
  observabilitySearchPath: string
  reconcileApiHint: Record<string, unknown>
  manualRetrySupported: boolean
}

export async function getAdminObservabilityToolCalls(
  executionId: string,
): Promise<ObservabilityToolCallsResponse> {
  const id = executionId.trim()
  if (!id) throw new AppError('executionId 不可为空', { status: 400 })
  try {
    return await httpClient
      .get(`${PREFIX}/executions/${encodeURIComponent(id)}/tool-calls`)
      .json<ObservabilityToolCallsResponse>()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

export async function getScenarioSkillScope(scenarioId: string): Promise<ScenarioSkillScopeResponse> {
  const sid = scenarioId.trim()
  if (!sid) throw new AppError('scenarioId 不可为空', { status: 400 })
  try {
    return await httpClient
      .get(`${PREFIX}/scenarios/${encodeURIComponent(sid)}/skill-scope`)
      .json<ScenarioSkillScopeResponse>()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

export async function getAdminObservabilityLlm(
  executionId: string,
): Promise<ObservabilityLlmUsageResponse> {
  const id = executionId.trim()
  if (!id) throw new AppError('executionId 不可为空', { status: 400 })
  try {
    return await httpClient
      .get(`${PREFIX}/executions/${encodeURIComponent(id)}/llm`)
      .json<ObservabilityLlmUsageResponse>()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

async function getExecutionSubResource<T>(executionId: string, suffix: string): Promise<T> {
  const id = executionId.trim()
  if (!id) throw new AppError('executionId 不可为空', { status: 400 })
  try {
    return await httpClient
      .get(`${PREFIX}/executions/${encodeURIComponent(id)}/${suffix}`)
      .json<T>()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

export function getAdminObservabilityExecutionQueue(
  executionId: string,
): Promise<ExecutionTaskQueueResponse> {
  return getExecutionSubResource<ExecutionTaskQueueResponse>(executionId, 'queue')
}

export function getAdminObservabilityExecutionEvents(
  executionId: string,
): Promise<ExecutionRuntimeEventsResponse> {
  return getExecutionSubResource<ExecutionRuntimeEventsResponse>(executionId, 'events')
}

export function getAdminObservabilityExecutionRetries(
  executionId: string,
): Promise<ExecutionRetriesResponse> {
  return getExecutionSubResource<ExecutionRetriesResponse>(executionId, 'retries')
}

export function getAdminObservabilityExecutionRecovery(
  executionId: string,
): Promise<ExecutionRecoveryResponse> {
  return getExecutionSubResource<ExecutionRecoveryResponse>(executionId, 'recovery')
}

File: server/chainup_agent/api/routers/admin_observability.py
"""Admin observability — persisted execution list/detail (`agent_execution`)."""

from __future__ import annotations

from datetime import datetime
from typing import Annotated

from chainup_agent.api.deps import DbSession, require_admin_console_bearer
from chainup_agent.api.schemas.admin_agent_executions import (
    AdminAgentExecutionItem,
    AdminAgentExecutionListResponse,
    admin_agent_execution_item_from_row,
)
from chainup_agent.api.schemas.admin_observability_timeline import (
    ObservabilityTimelineResponse,
    observability_timeline_item_from_row,
)
from chainup_agent.application.admin_agent_executions import (
    count_agent_executions_admin,
    delete_agent_execution_admin,
    get_agent_execution_admin,
    list_agent_executions_admin,
)
from chainup_agent.api.schemas.admin_observability_tool_llm import (
    LlmCallItem,
    LlmUsageSummaryResponse,
    ToolCallDetailItem,
    ToolCallsPageResponse,
)
from chainup_agent.application.admin_observability_execution_detail import (
    list_llm_usage_for_execution,
    list_tool_calls_for_execution,
)
from chainup_agent.application.admin_observability_execution_tabs import (
    build_execution_tabs_for_id,
    get_execution_recovery_view,
    list_execution_retries,
    list_execution_runtime_events,
    list_execution_task_queue,
)
from chainup_agent.api.schemas.admin_observability_execution_tabs import (
    ExecutionRecoveryResponse,
    ExecutionRetriesResponse,
    ExecutionRetryItem,
    ExecutionRuntimeEventItem,
    ExecutionRuntimeEventsResponse,
    ExecutionTaskQueueItem,
    ExecutionTaskQueueResponse,
)
from chainup_agent.api.schemas.admin_scenario_skill_scope import ScenarioSkillScopeResponse
from chainup_agent.application.agent_execution_events import list_timeline_events_for_execution
from chainup_agent.application.scenario_skill_scope import resolve_scenario_skill_scope_view
from chainup_agent.core.errors import AppError
from fastapi import APIRouter, Depends, Query
from fastapi.responses import Response

router = APIRouter(
    prefix="/api/v1/admin/observability",
    tags=["Admin — Observability"],
    dependencies=[Depends(require_admin_console_bearer)],
)

_ALLOWED_EXEC_STATES = frozenset({"ACCEPTED", "SUCCEEDED", "FAILED", "CANCELLED"})


def _optional_state_filter(raw: str | None) -> str | None:
    if raw is None or not str(raw).strip():
        return None
    s = str(raw).strip().upper()
    if s not in _ALLOWED_EXEC_STATES:
        raise AppError(
            code="VALIDATION_ERROR",
            message=("state 须为 ACCEPTED、SUCCEEDED、FAILED、CANCELLED 之一（或不传）"),
            status_code=422,
            details={"field": "state", "value": str(raw).strip()[:64]},
        )
    return s


@router.get(
    "/executions",
    response_model=AdminAgentExecutionListResponse,
    response_model_by_alias=True,
    summary="执行列表（`agent_execution` 分页 + 筛选）",
)
async def list_admin_observability_executions(
    db: DbSession,
    limit: int = Query(default=50, ge=1, le=200),
    offset: int = Query(default=0, ge=0),
    execution_id: str | None = Query(default=None, alias="executionId"),
    user_id: str | None = Query(default=None, alias="userId"),
    channel: str | None = None,
    scenario_id: str | None = Query(default=None, alias="scenarioId"),
    state: str | None = None,
    keyword: str | None = Query(
        default=None,
        description="模糊匹配 executionId / userId / scenarioId 子串",
        max_length=256,
    ),
    created_after: Annotated[datetime | None, Query(alias="createdAfter")] = None,
    created_before: Annotated[datetime | None, Query(alias="createdBefore")] = None,
) -> AdminAgentExecutionListResponse:
    st = _optional_state_filter(state)
    filt_eid = execution_id.strip() if execution_id and execution_id.strip() else None
    filt_uid = user_id.strip() if user_id and user_id.strip() else None
    filt_ch = channel.strip() if channel and channel.strip() else None
    filt_sid = scenario_id.strip() if scenario_id and scenario_id.strip() else None

    total = await count_agent_executions_admin(
        db,
        execution_id=filt_eid,
        user_id=filt_uid,
        channel=filt_ch,
        scenario_id=filt_sid,
        state=st,
        keyword=keyword,
        created_after=created_after,
        created_before=created_before,
    )
    rows = await list_agent_executions_admin(
        db,
        limit=limit,
        offset=offset,
        execution_id=filt_eid,
        user_id=filt_uid,
        channel=filt_ch,
        scenario_id=filt_sid,
        state=st,
        keyword=keyword,
        created_after=created_after,
        created_before=created_before,
    )
    items = [admin_agent_execution_item_from_row(r) for r in rows]
    return AdminAgentExecutionListResponse(items=items, total=total)


@router.get(
    "/scenarios/{scenario_id}/skill-scope",
    response_model=ScenarioSkillScopeResponse,
    response_model_by_alias=True,
    summary="场景 Skill 范围（治理面板只读 · SC-PM-21）",
)
async def get_admin_observability_scenario_skill_scope(
    scenario_id: str,
    db: DbSession,
) -> ScenarioSkillScopeResponse:
    sid = scenario_id.strip()
    if not sid:
        raise AppError(
            code="VALIDATION_ERROR",
            message="scenarioId 不能为空",
            status_code=422,
            details={"field": "scenarioId"},
        )
    doc = await resolve_scenario_skill_scope_view(db, scenario_id=sid)
    return ScenarioSkillScopeResponse.from_view(doc)


@router.get(
    "/executions/{execution_id}/timeline",
    response_model=ObservabilityTimelineResponse,
    response_model_by_alias=True,
    summary="执行时间线（FR-MC801 下限；从 Telegram 闪兑等写入 agent_execution_event）",
)
async def get_admin_observability_execution_timeline(
    execution_id: str,
    db: DbSession,
) -> ObservabilityTimelineResponse:
    parent = await get_agent_execution_admin(db, execution_public_id=execution_id)
    if parent is None:
        raise AppError(
            code="AGENT_ADMIN_EXECUTION_NOT_FOUND",
            message="未找到该 execution",
            status_code=404,
            details={"executionId": execution_id.strip()[:80]},
        )
    ev_rows = await list_timeline_events_for_execution(db, execution_public_id=execution_id)
    return ObservabilityTimelineResponse(
        items=[observability_timeline_item_from_row(r) for r in ev_rows]
    )


@router.get(
    "/executions/{execution_id}/tool-calls",
    response_model=ToolCallsPageResponse,
    response_model_by_alias=True,
    summary="工具调用明细 FR-MC803（trading.exchange_* · agent.tool.call）",
)
async def get_admin_observability_execution_tool_calls(
    execution_id: str,
    db: DbSession,
) -> ToolCallsPageResponse:
    parent = await get_agent_execution_admin(db, execution_public_id=execution_id)
    if parent is None:
        raise AppError(
            code="AGENT_ADMIN_EXECUTION_NOT_FOUND",
            message="未找到该 execution",
            status_code=404,
            details={"executionId": execution_id.strip()[:80]},
        )
    raw = await list_tool_calls_for_execution(db, execution_public_id=execution_id)
    return ToolCallsPageResponse(
        items=[ToolCallDetailItem.model_validate(x) for x in raw],
    )


async def _require_execution_or_404(db: DbSession, execution_id: str) -> None:
    parent = await get_agent_execution_admin(db, execution_public_id=execution_id)
    if parent is None:
        raise AppError(
            code="AGENT_ADMIN_EXECUTION_NOT_FOUND",
            message="未找到该 execution",
            status_code=404,
            details={"executionId": execution_id.strip()[:80]},
        )


@router.get(
    "/executions/{execution_id}/queue",
    response_model=ExecutionTaskQueueResponse,
    response_model_by_alias=True,
    summary="执行详情 · 任务队列（编排步骤 + 写路径未完成步）",
)
async def get_admin_observability_execution_queue(
    execution_id: str,
    db: DbSession,
) -> ExecutionTaskQueueResponse:
    await _require_execution_or_404(db, execution_id)
    row, events = await build_execution_tabs_for_id(db, execution_public_id=execution_id)
    assert row is not None
    raw = await list_execution_task_queue(db, execution=row, events=events)
    return ExecutionTaskQueueResponse(
        items=[ExecutionTaskQueueItem.model_validate(x) for x in raw],
    )


@router.get(
    "/executions/{execution_id}/events",
    response_model=ExecutionRuntimeEventsResponse,
    response_model_by_alias=True,
    summary="执行详情 · 运行事件摘录（tool/retry/对账等；全量检索走 FR-MC802）",
)
async def get_admin_observability_execution_events(
    execution_id: str,
    db: DbSession,
) -> ExecutionRuntimeEventsResponse:
    await _require_execution_or_404(db, execution_id)
    row, events = await build_execution_tabs_for_id(db, execution_public_id=execution_id)
    assert row is not None
    raw = await list_execution_runtime_events(db, execution=row, events=events)
    return ExecutionRuntimeEventsResponse(
        items=[ExecutionRuntimeEventItem.model_validate(x) for x in raw],
    )


@router.get(
    "/executions/{execution_id}/retries",
    response_model=ExecutionRetriesResponse,
    response_model_by_alias=True,
    summary="执行详情 · 重试历史（由时间线 retry/reconcile/unknown 摘录）",
)
async def get_admin_observability_execution_retries(
    execution_id: str,
    db: DbSession,
) -> ExecutionRetriesResponse:
    await _require_execution_or_404(db, execution_id)
    row, events = await build_execution_tabs_for_id(db, execution_public_id=execution_id)
    assert row is not None
    raw = await list_execution_retries(db, execution=row, events=events)
    return ExecutionRetriesResponse(
        retry_count=raw["retryCount"],
        items=[ExecutionRetryItem.model_validate(x) for x in raw["items"]],
    )


@router.get(
    "/executions/{execution_id}/recovery",
    response_model=ExecutionRecoveryResponse,
    response_model_by_alias=True,
    summary="执行详情 · 恢复与对账（对账状态 + 协查深链）",
)
async def get_admin_observability_execution_recovery(
    execution_id: str,
    db: DbSession,
) -> ExecutionRecoveryResponse:
    await _require_execution_or_404(db, execution_id)
    row, events = await build_execution_tabs_for_id(db, execution_public_id=execution_id)
    assert row is not None
    raw = await get_execution_recovery_view(db, execution=row, events=events)
    return ExecutionRecoveryResponse.model_validate(raw)


@router.get(
    "/executions/{execution_id}/llm",
    response_model=LlmUsageSummaryResponse,
    response_model_by_alias=True,
    summary="LLM 计量明细 FR-MC804（eventName llm.* · 无 messages 全文）",
)
async def get_admin_observability_execution_llm(
    execution_id: str,
    db: DbSession,
) -> LlmUsageSummaryResponse:
    parent = await get_agent_execution_admin(db, execution_public_id=execution_id)
    if parent is None:
        raise AppError(
            code="AGENT_ADMIN_EXECUTION_NOT_FOUND",
            message="未找到该 execution",
            status_code=404,
            details={"executionId": execution_id.strip()[:80]},
        )
    raw = await list_llm_usage_for_execution(db, execution_public_id=execution_id)
    calls = [LlmCallItem.model_validate(c) for c in raw.get("calls", [])]
    return LlmUsageSummaryResponse(
        model_id=raw.get("modelId"),
        input_tokens=raw.get("inputTokens"),
        output_tokens=raw.get("outputTokens"),
        total_tokens=raw.get("totalTokens"),
        calls=calls,
    )


@router.get(
    "/executions/{execution_id}",
    response_model=AdminAgentExecutionItem,
    response_model_by_alias=True,
    summary="执行详情（单条 `agent_execution`）",
)
async def get_admin_observability_execution(
    execution_id: str,
    db: DbSession,
) -> AdminAgentExecutionItem:
    row = await get_agent_execution_admin(db, execution_public_id=execution_id)
    if row is None:
        raise AppError(
            code="AGENT_ADMIN_EXECUTION_NOT_FOUND",
            message="未找到该 execution",
            status_code=404,
            details={"executionId": execution_id.strip()[:80]},
        )
    return admin_agent_execution_item_from_row(row)


@router.delete(
    "/executions/{execution_id}",
    status_code=204,
    summary="删除执行记录（硬删除 `agent_execution`，联调 / 运维清理）",
)
async def delete_admin_observability_execution(
    execution_id: str,
    db: DbSession,
) -> Response:
    deleted = await delete_agent_execution_admin(db, execution_public_id=execution_id)
    if not deleted:
        raise AppError(
            code="AGENT_ADMIN_EXECUTION_NOT_FOUND",
            message="未找到该 execution",
            status_code=404,
            details={"executionId": execution_id.strip()[:80]},
        )
    await db.commit()
    return Response(status_code=204)
